src/internal/time.ts

```diff
@@ -0,0 +1,110 @@
+import * as Async from '../Async'
+import { Fx, handle, ok } from '../Fx'
+import { Monotonic, Now, Sleep } from '../Time'
+
+export interface ScheduledTask {
+  readonly at: number
+  readonly task: () => void
+}
+
+export type SleepToAsync<E> = E extends Sleep ? Async.Async : never
+
+export class TimeStep {
+  private _monotonic = 0
+  private _target = 0
+  private _tasks: ScheduledTask[] = []
+  private _timeout: any
+  private _disposed = false
+
+  constructor(public readonly nowOrigin: number) { }
+
+  get monotonic(): number {
+    return this._monotonic
+  }
+
+  get now(): number {
+    return this.nowOrigin + this._monotonic
+  }
+
+  get target(): number {
+    return this._target
+  }
+
+  get disposed(): boolean {
+    return this._disposed
+  }
+
+  [Symbol.dispose](): void {
+    this._disposed = true
+    this._tasks = []
+    this.clearTimeout()
+  }
+
+  step(millis: number): Promise<void> {
+    if (this._disposed) return Promise.resolve()
+    this._target = this._target + Math.max(0, millis)
+    return this.runTasks()
+  }
+
+  waitAll(): Promise<void> {
+    return this.step(Infinity)
+  }
+
+  handle = <E, A>(f: Fx<E, A>): Fx<Exclude<E, Now | Monotonic | Sleep> | SleepToAsync<E>, A> => f.pipe(
+    handle(Now, () => ok(this.now)),
+    handle(Monotonic, () => ok(this.monotonic)),
+    handle(Sleep, ms => {
+      return Async.run(signal => new Promise<void>(resolve => {
+        const time = this._monotonic + Math.max(0, ms)
+        const t = { at: time, task: resolve }
+        this._tasks.push(t)
+        signal.addEventListener('abort', () => {
```

Comment:
If you add a `{once: true}` to the listener options, you don't even need the if-statement below since it's the only place it could possibly be removed

Reply:
Pretty keen observation.  Thanks 👍